08_irq/01_check_button/check_button.c

```diff
@@ -0,0 +1,163 @@
+// SPDX-License-Identifier: GPL-2.0
+#define pr_fmt(fmt) KBUILD_MODNAME ": %s: " fmt, __func__
+
+#include <linux/init.h>
+#include <linux/module.h>
+#include <linux/kernel.h>
+#include <linux/gpio.h>
+#include <linux/timer.h>
+#include <linux/jiffies.h>
+#include <linux/hrtimer.h>
+#include <linux/ktime.h>
+#include <linux/mic_bus.h>
+
+MODULE_LICENSE("GPL");
+MODULE_AUTHOR("Oleksiy Lyubochko <oleksiy.m.lyubochko@globallogic.com>");
+MODULE_DESCRIPTION("A simple example Linux module.");
+MODULE_VERSION("0.01");
+
+#define GPIO_NUMBER(port, bit) (32 * (port) + (bit))
+
+/* On-board LESs
+ *  0: D2	GPIO1_21	heartbeat
+ *  1: D3	GPIO1_22	uSD access
+ *  2: D4	GPIO1_23	active
+ *  3: D5	GPIO1_24	eMMC access
+ *
+ * Note that uSD and eMMC LEDs are not used in nfs boot mode, but they are
+ * already requested by correspondent drivers.
+ * So that, we don't use gpio_request()/gpio_free() here, but control these
+ * LEds (dirty!).
+ * Use request/free if you want to control free GPIO routed to board
+ * connectors.
+ */
+
+#define LED_SD  GPIO_NUMBER(1, 22)
+#define LED_MMC GPIO_NUMBER(1, 24)
+
+#define BUTTON  GPIO_NUMBER(2, 8)
+#define GPIO_LED 49
+#define GPIO_BUTTON 117
+#define TIMER_DELAY 20
+
+static int button_state;
+static struct timer_list simple_timer;
+static struct hrtimer hr_timer;
+static unsigned long timer_interval_ns = 1e6;
+
+static enum hrtimer_restart hrtimer_handler(struct hrtimer *timer)
+{
+	gpio_direction_output(GPIO_LED, 0);
+	pr_info("HR Timer module calback\n");
+
+	return HRTIMER_NORESTART;
+}
+
+void timer_handler(struct timer_list *timer)
+{
+	int state;
+	ktime_t interval;
+
+	mod_timer(timer, jiffies + (msecs_to_jiffies(TIMER_DELAY)));
+
+	state = gpio_get_value(GPIO_BUTTON);
+
+	if (unlikely(button_state != state)) {
+		pr_info("Timerfunction, button state have changed\n");
+		gpio_direction_output(LED_MMC, state);
+
+		gpio_direction_output(GPIO_LED, 1);
```

Comment:
gpio_direction_output should be used only during initialization
during runtime use gpio_set_value